Refactor to preserve current behaviour but add observer based callbacks

guyboertje opened the pull request:
Maybe this can serve as a base for time based unwatch

jordansissel commented on lib/filewatch/tail.rb:
```diff
@@ -20,6 +26,13 @@ class Tail
+    alias_method :subscribe, :yield_subscribe
+
+    def self.new_observing(opts = {})
+      alias_method :subscribe, :observe_subscribe
```
This feels... unsafe? Modifying the instance methods (of all instances?) Did you mean to use define_singleton_method ?

guyboertje replied:
Yeah, I'm not that happy with this.  I will rethink it.

guyboertje commented:
@jordansissel - made some changes:
- Refactored the switch between tailing and observing.
- added some timeout code - not altogether complete.  Need to write pre-expired files to sincedb and not only ignore them.

jordansissel commented on lib/filewatch/watch.rb:
```diff
@@ -18,7 +45,7 @@ def initialize(opts={})
       @watching = []
       @exclude = []
-      @files = Hash.new { |h, k| h[k] = Hash.new }
+      @files = Hash.new { |h, k| h[k] = WatchedFile.new(nil, false) }
```
+1 having state kept in something firmer than a Hash. :+1:

guyboertje replied:
Yeah and with predicates that read better.  Also, it holds the path of the file internally and not just in the `@files.keys`.

It also opens up other possibilities:
- passing it as the yield arg to Tail
- giving it the responsibility of opening and closing its file
- retrieving the Stat.
- making a WatchedFileWindows and WatchedFileNix variants that know what to do in a platform related way.
- maybe removes the necessity to memoize the opened file in the Tail classes

jordansissel commented on lib/filewatch/watch.rb:
```diff
@@ -108,23 +130,34 @@ def each(&block)
-          @files[path][:size] = stat.size
-          @files[path][:inode] = inode
-        end # @files.keys.each
+          watched_value.size = stat.size
```
Idea for this WatchedFile - what about using watched_value.update(stat) ? Maybe not... I'm mostly feeling a dislike of accessors lately, what do you think? Open to also deferring any changes based on this comment to a future PR

guyboertje replied:
I like the idea and will add it.

jordansissel commented on lib/filewatch/watch.rb:
```diff
@@ -155,6 +188,21 @@ def subscribe(stat_interval = 1, discover_interval = 5, &block)
+    end
+
+    def file_expired?(stat)
+      Time.now.to_i > (stat.mtime.to_i + @ignore_after)
```
I think you can subtract Time values, so maybe this makes more sense:

```
(Time.now - stat.mtime) > @ignore_after
```

I think it should work the same as your code, but avoids .to_i and compares directly to the duration (ignore_after) and might make it seem easier to read?

Alternately, a new method `file_age(stat)` that returns the `Time.now - stat.mtime` and then this method becomes `file_age(stat) > @ignore_after` 

(this is mostly trivial style concerns, so we can defer/delay resolution of this.)

guyboertje replied:
I am always a bit wary of operations on Time objects especially when there are hidden operations that could involve timezones (I have been bitten sooo many times by timezone issues).

However, just now, I looked at the JRuby code (I do this quite often to see exactly what is happening under the hood) and in RubyTime the `opMinus` method converts LHS and RHS to ints then floats before subtraction and returns a float. Then, in RubyFloat the `op_gt` method converts the RHS to a Java double before comparing it to the RubyFloat's internal double value.

Phew, a lot of conversions, I think using integers is more performant.
But I agree that the line is a bit WTF and violates POLS.

jordansissel commented on lib/filewatch/tail_base.rb:
```diff
@@ -0,0 +1,217 @@
+
+    public
+    def initialize(opts={})
+      @iswindows = ((RbConfig::CONFIG['host_os'] =~ /mswin|mingw|cygwin/) != nil)
```
This can probably be a constant.

jordansissel replied:
If this was just copied/moved code, please disregard. :)

guyboertje replied:
Renamed tail.rb to tail_base.rb and modified to be a module.

jordansissel commented:
Feeling pretty good so far about this PR, @guyboertje. Tests are passing for me :+1:

electrical commented:
Haven't run any tests or anything but code looks sane.

jsvd commented on lib/filewatch/watch.rb:
```diff
@@ -86,18 +126,18 @@ def inode(path,stat)
-        @files.keys.each do |path|
-          if ! @files[path][:create_sent]
-            if @files[path][:initial]
+        @files.each do |path, watched_value|
```
I suggest renaming `watched_value` to `watched_file`

guyboertje replied:
I wanted to keep the notion of Value object, however that is not strongly communicated in the name `watch_value`.  I will go with watched_file

jsvd commented on lib/filewatch/watch.rb:
```diff
@@ -201,9 +279,19 @@ def synchronized(&block)
+    end
+
+    private
+    def unquit
```
maybe abort_quit instead?

guyboertje replied:
how about `reset_quit`?

jsvd replied:
:+1:

jsvd commented:
code LGTM and tests run (though the take a LONG time to finish: `Finished in 52.38 seconds`)

jordansissel commented:
Your newest two commits LGTM.

LGTM for merging.
